test/functional/horizon_controller_test.go

```diff
@@ -0,0 +1,279 @@
+package functional_test
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/google/uuid"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	horizon "github.com/openstack-k8s-operators/horizon-operator/api/v1alpha1"
+	keystonev1 "github.com/openstack-k8s-operators/keystone-operator/api/v1beta1"
+	condition "github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+)
+
+const (
+	// TODO(gibi): Do we want to test in a realistic namespace like "openstack"?
+	TestNamespace = "default"
```

Comment:
one hint. In envtest you cannot really delete a namespace https://book.kubebuilder.io/reference/envtest.html#namespace-usage-limitation but probably you want to isolate each test case from each other so if one leaves some resources in the namespace it does not effect the later test running in that namespace. So we started using a generated namespace name for each test to isolate them.